ALSpringBackend/src/main/java/com/ardaslegends/data/service/Pathfinder.java

```diff
@@ -0,0 +1,156 @@
+    //smallest weights between startRegion and all the other nodes
+    Map<Region, Integer> smallestWeights = new HashMap<>();
+    //for convenience, mark distance from startRegion to itself as 0
+    smallestWeights.put(startRegion, 0);
+
+    //implicit graph of all nodes and previous node in ideal paths
+    Map<Region, Region> prevNodes = new HashMap<>();
+
+    //use queue for breadth first search
+    //for convenience, we'll use an array, but linked list would be preferred
+    ArrayList<Region> nodesToVisitQueue= new ArrayList<>();
+
+    //record visited nodes with a set. The string is the toString of a HexID
+    Set<String> visitedNodes = new HashSet<>();
+    visitedNodes.add(startRegion.getId());
+
+    Region currentNode = startRegion;
+    //loop through nodes
+    while (true) {
+      //get the shortest path so far from start to currentNode
+      final int dist = smallestWeights.get(currentNode);
+
+      //iterate over current child's nodes and process
+      Set<Region> neighbourRegions = currentNode.getNeighboringRegions();
+      for (Region neighbourRegion : neighbourRegions) {
+        //add node to queue if not already visited
+        if (
+          !visitedNodes.contains(neighbourRegion.getId()) &&
+          !nodesToVisitQueue.contains(neighbourRegion)
+        ) {
+          nodesToVisitQueue.add(neighbourRegion);
+        }
+
+        // CALCULATE COST
+        int thisDist = 0;
+
+        // Check if we can move to that region as an army
+        if (player.getRpChar().getBoundTo() != null || isLeaderMove) {
+          if (player.getFaction().getAllies().stream().noneMatch(faction ->
+                  neighbourRegion.getClaimedBy().contains(faction))
+                  && !neighbourRegion.getClaimedBy().contains(player.getFaction())
+                  && !neighbourRegion.getClaimedBy().isEmpty()) {
+            thisDist = 1000;
+          }
+        }
+
+        // Check if we are embarking or disembarking
+        if (currentNode.getRegionType() != RegionType.SEA && neighbourRegion.getRegionType() == RegionType.SEA) {
+          boolean canEmbark = false;
+          for (ClaimBuild claimbuild : currentNode.getClaimBuilds()) {
+            if (claimbuild.getSpecialBuildings().contains(SpecialBuilding.HARBOUR)) {
+              thisDist++;
+              canEmbark = true;
+              break;
+            }
+          }
+          // It's a code smell I know. Had to find a quick fix.
+          // Should be part of the next code review.
+          if (!canEmbark) {
+            thisDist = 1000;
+          }
+        } else if (currentNode.getRegionType() == RegionType.SEA && neighbourRegion.getRegionType() != RegionType.SEA) {
+          thisDist += dist + neighbourRegion.getCost() + 1;
+        } else {
+          thisDist += dist + neighbourRegion.getCost();
+          // Check if there is no army bound to character
+          if (player.getRpChar().getBoundTo() == null && !isLeaderMove) {

```

Comment:
If my way of thinking is correct (idk tbh, very hard to review algorithm code), then this isn't very optimal, since we don't really know if this is an army movement or not. The way I understand it, is that you want to check here if the movement is a RpChar movement and therefore have double movement speed. I'd rather have a parameter for this function, something like `isArmyMovement` and let the method that calls the pathfinder handle all the checking etc.

Again, I'd say that all validation should be handled outside the Pathfinder. Things like checking if the movement is a leaderMove and if the army can be moved by the player should be handled by the Service that calls the Pathfinder. The Service checks if the player is allowed to move the army (and also checks every other business logic), and if the move is legal, the last thing that is done is calling the pathfinder to calculate the cost.

If we go by this logic, leader moves should be completely irrelevant for the Pathfinder, right? Since leaders don't move faster or something, their movements are just regular movements

Reply:
# VERY IMPORTANT BTW

Reply:
True concerning leader movements. My logic when I quickly added the boolean, was that a player might move but he is not bound to an army (thus specifying that he is moving an army he is not bound to). Which is indeed irrelevant information for the pathfinder and he just needs to know if it's an rp char or army moving. This can be easily optimized in the last lines when I have the total cost, eventually dividing it by checking the bool

Reply:
So we removing the `isLeaderMove` boolean and get a `isRpcharMove` boolean instead?

Reply:
nvm it already happened

Reply:
Think this conversation is resolved right?

Reply:
Yep